tinygrad/renderer/cstyle.py

```diff
@@ -200,9 +201,18 @@ class ClangRenderer(CStyleLanguage):
   type_map = {dtypes.bool:"_Bool", dtypes.half:"__fp16"}
   code_for_op = {**CStyleLanguage().code_for_op, BinaryOps.MAX: lambda a,b,dtype: f"(({a}>{b})?{a}:{b})"}
 
+  if AMX:
+    tc_types = [(dtype, amx_size//dtype.itemsize) for dtype, amx_size in zip([dtypes.float], [64])]
+    tensor_cores = [TensorCore(dims=(sz,sz,sz), threads=[(0,sz),(1,sz)], dtype_in=dtype, dtype_out=dtype) for dtype, sz in tc_types]
+
   def render_kernel(self, function_name, kernel, bufs, uops, prefix=None) -> str:
-    prefix = [_make_clang_dtype(self, dtype) for dtype in dedup(uop.dtype for uop in uops if uop.dtype is not None and uop.dtype.count>1)]
-    return super().render_kernel(function_name, kernel, bufs, uops, prefix)
+    prefix, macros = [_make_clang_dtype(self, dtype) for dtype in dedup(uop.dtype for uop in uops if uop.dtype is not None and uop.dtype.count>1)], []
+    for name, (N, M, K), dtype_in, _, _, _, _ in dedup([uop.arg for uop in uops if uop.op is UOps.WMMA]):
+      macros = ['#define AMX_SET(imm5) __asm("nop\\nnop\\nnop\\n.word (0x201000+(%0<<5)+%1)" : : "i"(17), "i"(imm5) : "memory")', '#define AMX(op, gpr, btf) __asm(".word (0x201000+(%0 << 5)+0%1-((0%1>>4)*6))" : : "i"(op), "r"((unsigned long long)(gpr)+(btf)) : "memory")'] # noqa: E501
+      prefix += [f"""{dtype_in.vec(K*K).name} __{name}({dtype_in.vec(N).name} data1, {dtype_in.vec(M).name} data2){{
+  AMX_SET(0);\n  AMX(0, (int *)(&data2), 0ull<<62); AMX(1, (int *)(&data1), 0ull<<62); AMX(12, 0, 0ull); {dtype_in.vec(K*K).name} data = {{0}};
+  for(int ridx0 = 0; ridx0 < 16; ridx0++){{ AMX(5, (int *)(&data), 0ull<<62 | (ridx0*4ull)<<56 | ridx0*64ull); }}\n  AMX_SET(1);\n  return data;\n}}"""] # noqa: E501
```

Comment:
Oh, this loop is why it's slow! It should be used as MAC array, not as a matrix core. (also, is 16 right for float16?)

Reply:
Sorry I don't understand what you meant by MAC array vs matrix core. Right now no accumulation is being carried out in AMX reg.

No, 32 for half and 8 for double. 

Other types support are pending though, generalizing the loop. AMX code op also change based on typing. Would you like to add support for those types in this pr? For double more things might need to change as it does not vectorizes, but for half it should be straightforward.

Reply:
Based on tests I made, what has the biggest impact on performance is using pointers for the structs in both ways as it avoid coping (for float it copies float256 for every iteration) and doing the accumulation in AMX itself.

Reply:
Why is there a loop here? It should be a single call to the AMX

Reply:
The following figure represents how the AMX registers are being used. For this first support, only one register is used for x and y (`x0` and `y0`). AMX fma opeartion between `x0` and `y0` is basically a full outer product with accumulation with what is on the `z registers`. This full outer product is distributed among the colored `z registers`. AMX does not support storing all those `z registers` all at once, they have to be stored individually, hence the loop, every iteration stores one reg. This loop can be unrolled though I believe for better performance 

![image](https://github.com/user-attachments/assets/da25d288-66b0-4bfe-b432-a1feb66ea3d7)

Reply:
For full f32 performance, double loads are used for x and y and four full outer products are performed, accumulating on the remaining z registers